### ISSUE TITLE
State Khronos' official stance on enum groupings

### ISSUE THREAD

**Issue by Perksey** [KhronosGroup/OpenGL-Registry#280]
Enum groupings are great for downstream consumers of the Khronos specification (i.e. bindings generators for other languages such as C#) and are depended on by projects such as OpenTK. However, Khronos seems to be neglecting these groupings, breaking code generators that depend on them. I feel that Khronos either needs to stop neglecting these groupings or officially state that enum groups are deprecated so that library developers can start looking for alternatives.

Regards,

Dylan.

**Comment by oddhack** [comment]
The enum groups are something left over from SGI's 25 year old gl.spec file format, and I imported them as they were in the GL 1.2 spec files when we moved to the gl.xml format. Khronos makes exactly no use of them for our purposes, and we don't have many cycles free for working on this stuff. So in that sense they have always been deprecated.

I'll take another look at #274 soon, which is I assume what you're referring to here. I did see your comments, but the PR creator hasn't responded there, and I was hoping to get their signoff before proceeding.

**Comment by Perksey** [comment]
This is somewhat related to that PR yes, but I wanted to get Khronos' general stance as I don't want to continue working on improving the spec's groupings if they're not even supported by Khronos. Currently, our C# bindings to OpenGL are stuck on specification version `681c365c` because our binder can't generate strongly-typed enums of enum groups that are referenced by functions but don't exist. Moreover, a lot of these enum groups are incorrect or incomplete.

I'm not trying to get you to prioritise #274, I'm trying to get support for enum groups picked up at least a little bit, or have them dropped entirely.

**Comment by Perksey** [comment]
Any updates here? I think that Khronos needs to decide whether to keep these groupings or not, as to eliminate downstream consumers of the spec depending on these broken groupings. In an ideal world, Khronos would actively try to encourage vendors to keep these groupings up-to-date, but I understand that not everyone has time to do so.

It's a case of mend or remove, in my opinion. 

I understand the Khronos Group might be hesitant to remove these groupings from the specification, because yeah a lot of downstream consumers use them, myself included. But if you were worried about these consumers, surely you'd fix the groupings - downstream consumers already can't use the groupings in their current state, so at this point removing them makes no difference to the state of them.

**Comment by pdaniell-nv** [comment]
@Perksey and @oddhack do you have a sense of how much effort it would take to fix the group definitions?

**Comment by Perksey** [comment]
@pdaniell-nv To fix the enum groupings, you need to have a relatively good grasp of what constants are actually applicable for a given function/extension.

As an external contributor, this is extremely hard as @frederikja163 and I spent a solid 2 hours reading the text extension proposals/specifications, and then adding them into the spec manually. All this, with little knowledge of the extensions to which the groupings regard, will take a long time.

However, let's say you work at NVIDIA and are very familiar with how the extension functions work and what `GLenum` values they take, then it'll take no time at all. So what I could suggest is vendor developers volunteer to fix the groupings for their extensions (i.e. @pdaniell-nv does the NVIDIA functions, etc). 

Realistically, it's a simple job it just requires extensive knowledge of the extensions and functions within OpenGL. Which is probably why it took Fred and I so long as we kept having to read up on what the functions actually did.

> **TL;DR**: All you need is someone from each vendor to look over their extensions and create groups for each `GLenum` parameter. You just need to know what values the functions take.

**Comment by oddhack** [comment]
I missed replying earlier, but TBH I don't think we have a big advantage over anyone outside Khronos in terms of updating the enum groups. The information isn't, by and large, automatically extractable from the core spec or the registry extension specs - and in many cases, the extension authors have long since moved on to other things. So it comes down to a lot of painstaking review as @Perksey notes. It is not a complex job, but it is a very tedious and time-consuming one. NVIDIA would take the biggest hit doing this since they've authored far more extensions than anyone else.

The disincentive to removing them is that we don't, and can't know who the downstream consumers are unless they speak up here. Any time we remove some content like this, we end up getting complaints from projects we'd never even heard of previously. So I would not want to do that. We might add a comment to gl.xml noting the enum groups are not necc. up to date, or something of that nature.

**Comment by Perksey** [comment]
Thanks for clearing that up. Personally I don't like the idea of keeping in broken groupings, but I get the reasoning behind it. If we could officially deprecate the groupings in some way though, that'd be great.

**Comment by pdaniell-nv** [comment]
We discussed this again in the OpenGL/ES meeting today.

> We might add a comment to gl.xml noting the enum groups are not necc. up to date, or something of that nature.
@oddhack can your propose something along these lines. Thanks.

**Comment by realvictorprm** [comment]
This is a huge mess and it's incredible sad that the lazyness to make a single small addition in a documentation file created this over time.

**Comment by realvictorprm** [comment]
Wouldn't it be possible to look the enum groupings up in the human specification/documentation for a OpenGL function?

**Comment by oddhack** [comment]
Yes, that's what would be required. With the proviso that you would have to look up every extension that may interact with that function, as well as the core API spec.

**Comment by oddhack** [comment]
@pdaniell-nv #291

**Comment by realvictorprm** [comment]
ok thank you :)
